/*
 * Copyright (c) Facebook, Inc. and its affiliates.
 * All rights reserved.
 *
 * This source code is licensed under the BSD-style license found in the
 * LICENSE file in the root directory of this source tree.
 */

syntax = "proto3";

package tensorpipe.proto;

message SpontaneousConnection {}

message RequestedConnection {
  uint64 registration_id = 1;
}

enum DeviceType {
  DEVICE_TYPE_UNSPECIFIED = 0;
  DEVICE_TYPE_CPU = 1;
}

message MessageDescriptor {
  message TensorDescriptor {
    DeviceType device_type = 1;
    int64 size_in_bytes = 2;

    string side_channel_name = 3;
    bytes side_channel_data = 4;

    bytes user_data = 5;
  }

  int64 size_in_bytes = 1;
  repeated TensorDescriptor tensor_descriptors = 2;
}

message Request {}

message Message {
  message Tensor {
    bytes data = 1;
  }

  bytes data = 1;
  repeated Tensor tensors = 2;
}

message Packet {
  oneof type {
    SpontaneousConnection spontaneous_connection = 1;
    RequestedConnection requested_connection = 2;
    MessageDescriptor message_descriptor = 3;
    Request request = 4;
    Message message = 5;
  }
}

message BasicChannelOperation {
  uint64 operation_id = 1;
  uint64 size_in_bytes = 2;
}

message BasicChannelPacket {
  oneof type {
    BasicChannelOperation request = 1;
    BasicChannelOperation reply = 2;
  }
}
